src/coreComponents/physicsSolvers/fluidFlow/wells/CompositionalMultiphaseWell.cpp

```diff
@@ -757,7 +760,7 @@ void CompositionalMultiphaseWell::updateVolRatesForConstraint( WellElementSubReg
       real64 const currentTotalRate = connRate[iwelemRef];
 
       // Step 2.1: compute the inverse of the total density and derivatives
-
+      massDensity =totalDens[iwelemRef][0]; // need to verify this is surface dens
```

Comment:
looks like it's surface density when `useSurfaceConditions=true`
i think it will also work when `useSurfaceConditions=false` but may require derivatives for `massDensity`?
did you consider storing directly `currentTotalRate` instead of dealing with density?

Reply:
My intent was for option valid with surfacecond=true... I will add code to report error otherwise..Also if false the code needs to report surface and res rates for option to be of value.   I could be wrong but usesurfacecond=false is pressure at top seg, not reservoir conditions. I've never seen this , for example voidage replacement constraints are at res conditions...     In terms of total rate, I setup mass constraint to support "surfaceshutin w/crossflow" in the future, if that was the question..

Reply:
sounds good, thank you for the clarifications

> I could be wrong but usesurfacecond=false is pressure at top seg, not reservoir conditions.

I thinks it's correct yes

> I've never seen this , for example voidage replacement constraints are at res conditions...

I always thought that this reservoir conditions option is mostly for the developers to debug and test but I might be wrong

> In terms of total rate, I setup mass constraint to support "surfaceshutin w/crossflow" in the future, if that was the question..

Ok got it